File: Algorithms/0377_Combination_Sum_IV/Java/CombinationSumIvSolution2.java

// Space: O(n)
// Time: O(n)

public class Solution {

    public int combinationSum4(int[] nums, int target) {
        Arrays.sort(nums);
        if (nums[0] > target) return 0;

        int[] dp = new int[target + 1];
        for (int num : nums) {
            if (num < dp.length) dp[num] = 1;
        }

        for (int i = 1; i < dp.length; i++) {
            for (int num : nums) {
                if (i + num <= target) dp[i + num] += dp[i];
            }
        }
        return dp[dp.length - 1];
    }
}

File: Algorithms/0377_Combination_Sum_IV/Python/Combination_Sum_IV_Solution_2.py

# Space: O(n)
# Time: O(n^2)

class Solution:
    def combinationSum4(self, nums, target):
        nums = sorted(nums)
        if nums[0] > target: return 0

        dp = [0 for _ in range(target + 1)]
        for num in nums:
            if num < len(dp):
                dp[num] = 1

        for i in range(1, len(dp)):
            for num in nums:
                if i + num <= target:
                    dp[i + num] += dp[i]

        return dp[-1]




